### PR TITLE
Create a map hammer

### PR DESCRIPTION
**Pull request by daviddrysdale**

### CODE REVIEW

File: integration/maptest/hammer.go [hunk 1]
```diff
@@ -0,0 +1,678 @@
+// Copyright 2017 Google Inc. All Rights Reserved.
```
Comment (gdbelvin) [inline integration/maptest/hammer.go]: This looks like it's setting up various `hammer` objects.  Perhaps this should live in a `hammer` package?  Perhaps `testonly/hammer`? As it is `maptest` holds unittests exclusively.

File: integration/maptest/hammer.go [hunk 2]
```diff
@@ -0,0 +1,678 @@
...
+	// How far beyond current revision to request for invalid requests
+	invalidStretch = int64(10000)
+)
+
+var (
+	maxRetryDuration = 60 * time.Second
+)
+
+var (
+	// Metrics are all per-log (label "logid"), but may also be
```
Comment (AlCutter) [inline integration/maptest/hammer.go]: per map/ "mapid"

File: integration/maptest/hammer.go [hunk 3]
```diff
@@ -0,0 +1,678 @@
...
+
+// HammerMap performs load/stress operations according to given config.
+func HammerMap(cfg HammerConfig) error {
+	ctx := context.Background()
+
+	s := newHammerState(&cfg)
+
+	ticker := time.NewTicker(cfg.EmitInterval)
+	go func(c <-chan time.Time) {
+		for _ = range c {
```
Comment (AlCutter) [inline integration/maptest/hammer.go]: `for range c {}` :)

File: integration/maptest/hammer.go [hunk 4]
```diff
@@ -0,0 +1,678 @@
...
+		return -1, false
+	}
+	return rand.Intn(i), true
+}
+
+func (s *hammerState) updateContents(rev int64, leaves []*trillian.MapLeaf) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	// Shuffle earlier contents along.
```
Comment (AlCutter) [inline integration/maptest/hammer.go]: I wonder would using `container/list` or perhaps `container/ring` help?

File: integration/maptest/maphammer/main.go [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+
+func main() {
+	flag.Parse()
+	if *mapIDsFlag == "" {
+		glog.Exit("Test aborted as no map IDs provided (via --map_ids)")
+	}
+	if *seed == -1 {
+		*seed = time.Now().UTC().UnixNano() & 0xFFFFFFFF
+	}
+	fmt.Printf("Today's test has been brought to you by the letters M, A and P and the number %#x\n", *seed)
```
Comment (AlCutter) [inline integration/maptest/maphammer/main.go]: missing Oxford comma :)

File: integration/maptest/hammer.go (hunk 1 above)
Comment (daviddrysdale) [reply integration/maptest/hammer.go]: I'm hoping to use some of the `maptest` utility functions for verification (e.g. `verifyGetMapLeavesResponse` looks potentially useful).  But any such utilities should be promoted to be publicly visible and then the hammer code can live on its own.

Comment (gdbelvin) [reply integration/maptest/hammer.go]: Hmm, good point. I think this is a hint that we need something of the form of `client/mapverifier` which all clients can use.

File: integration/maptest/hammer.go (hunk 4 above)
Comment (daviddrysdale) [reply integration/maptest/hammer.go]: Not obviously.  `container/list` has no random access, which we use a bunch, and `container/ring` doesn't have the concept of a head, which we need for the 'most-recent' entries.

**Comment by daviddrysdale** [comment]
PTAL (commits rebased and updated so I could move the code into `testonly/hammer`)

File: integration/maptest/hammer.go (hunk 2 above)
Comment (daviddrysdale) [reply integration/maptest/hammer.go]: Done

File: integration/maptest/hammer.go (hunk 3 above)
Comment (daviddrysdale) [reply integration/maptest/hammer.go]: Done

File: integration/maptest/maphammer/main.go (hunk 5 above)
Comment (daviddrysdale) [reply integration/maptest/maphammer/main.go]: Done

**Comment by AlCutter** [comment]
"The job exceeded the maximum time limit for jobs, and has been terminated." :(

**Comment by daviddrysdale** [comment]
Yeah, given the current state of Travis I'll probably drop the last commit in the sequence.

However, given that it's not particularly fast on local machines either, we could probably do with some profiling of the map stuff...

**Pull request merged**